src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/VisualStudioSolutionCloseChangeTrigger.cs

```diff
@@ -0,0 +1,111 @@
+    {
+        private IVsSolution? _solution;
+        private readonly IServiceProvider _serviceProvider;
+        private readonly JoinableTaskContext _joinableTaskContext;
+
+        private uint _cookie;
+        private ProjectSnapshotManagerBase? _projectSnapshotManager;
+
+        [ImportingConstructor]
+        public VisualStudioSolutionCloseChangeTrigger(
+           [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
+           JoinableTaskContext joinableTaskContext)
+        {
+            _serviceProvider = serviceProvider;
+            _joinableTaskContext = joinableTaskContext;
+        }
+
+        public override void Initialize(ProjectSnapshotManagerBase projectManager)
+        {
+            _projectSnapshotManager = projectManager;
+
+            _ = _joinableTaskContext.Factory.RunAsync(async () =>
+            {
+                await _joinableTaskContext.Factory.SwitchToMainThreadAsync();
+
+                if (_serviceProvider.GetService(typeof(SVsSolution)) is IVsSolution solution)
+                {
+                    _solution = solution;
+                    _solution.AdviseSolutionEvents(this, out _cookie);
+                }
+            });
+        }
+
+        public void Dispose()
+        {
+            _solution?.UnadviseSolutionEvents(_cookie);
+        }
+
+        public int OnAfterOpenProject(IVsHierarchy pHierarchy, int fAdded)
+        {
+            return HResult.NotImplemented;
+        }
+
+        public int OnQueryCloseProject(IVsHierarchy pHierarchy, int fRemoving, ref int pfCancel)
+        {
+            return HResult.NotImplemented;
+        }
+
+        public int OnBeforeCloseProject(IVsHierarchy pHierarchy, int fRemoved)
+        {
+            return HResult.NotImplemented;
+        }
+
+        public int OnAfterLoadProject(IVsHierarchy pStubHierarchy, IVsHierarchy pRealHierarchy)
+        {
+            return HResult.NotImplemented;
+        }
+
+        public int OnQueryUnloadProject(IVsHierarchy pRealHierarchy, ref int pfCancel)
+        {
+            return HResult.NotImplemented;
+        }
+
+        public int OnBeforeUnloadProject(IVsHierarchy pRealHierarchy, IVsHierarchy pStubHierarchy)
+        {
+            return HResult.NotImplemented;
+        }
+
+        public int OnAfterOpenSolution(object pUnkReserved, int fNewSolution)
+        {
+            _projectSnapshotManager?.SolutionOpened();
+            return VSConstants.S_OK;
+        }
+
+        public int OnQueryCloseSolution(object pUnkReserved, ref int pfCancel)
+        {
+            return HResult.NotImplemented;
+        }
+
+        public int OnBeforeCloseSolution(object pUnkReserved)
+        {
+            _projectSnapshotManager?.SolutionClosed();
+            return VSConstants.S_OK;
+        }
+
+        public int OnAfterCloseSolution(object pUnkReserved)
+        {
+            _projectSnapshotManager?.SolutionOpened();
```

Comment:
Hmmm, I wonder if we need to leave our project snapshot manager open for business when VS has 0 solutions active. I don't feel strongly either way but it's just an interesting detail 😄